Demystifying Python (#113)
arcus/education_modules: intro_to_python/intro_to_python.md

author:   Meredith Lee
email:    leemc@chop.edu
version:  2.0.0
module_template_version: 2.0.0
language: en
narrator: UK English Female
title: Introduction to Python
comment:  This module introduces the Python programming language, explores why Python is useful in research, and describes how to download Python and Jupyter.
long_description: Python is a versatile programming language that is frequently used for data analysis, machine learning, web development, and more. If you are interested in learning about any of these topics and are new to programming, are joining a team where Python is used, or have just heard about Python and are looking for a general introduction to the language, this module is a good place to start. This is appropriate for someone at the beginner level.
estimated_time: 20m
@learning_objectives  

After completion of this module, learners will be able to:

- Describe what Python is and why they might want to use it for research
- Identify several ways to write Python code
- Understand the purpose and utility of a Jupyter notebook

> - Download Python to their own computers

Reviewer: this needs to be updated to either also include google colab info on the same objective or add it as a fifth objective 
(QA item: "All major topics covered are represented by a learning objective -- there's no significant knowledge imparted that isn't specified in the learning objectives" )

Reviewer: actually, looking back at the formative assessment section, every LO needs to have at least one Q of its own, so I think that download python/colab would need to be a singular LO unless you want to add another Q (which tbh seems like overkill to me)

Reviewer: actually, looking back at the formative assessment section, every LO needs to have at least one Q of its own, so I think that download python/colab would need to be a singular LO unless you want to add another Q (which tbh seems like overkill to me)

> - Download Python to their own computers

Reviewer: this needs to be updated to either also include google colab info on the same objective or add it as a fifth objective 
(QA item: "All major topics covered are represented by a learning objective -- there's no significant knowledge imparted that isn't specified in the learning objectives" )

author:   Meredith Lee
email:    leemc@chop.edu

> version:  2.0.0

Reviewer: ```suggestion
version:  1.0.0
```
since we're now entirely renaming + relocating the module, do we think we should go back to a 1.0.0 version?

Reviewer: For a mom,ent I thought this may have been covered in the versioning guidelines, but I just looked back and it doesn't look like it was (specifically how to handle when the rescoping involves renaming). I'm leaning toward doing 1.0.0 for this, but haven't thought on it particularly hard so feel free to push back!

module_template_version: 2.0.0
language: en
narrator: UK English Female

> title: Introduction to Python

Reviewer: Should we consider renaming this module getting started with python, getting set up with python or something like that? (I know annoyingly that would require creating new folders and lots of renaming stuff)

I'm just wondering if Introduction to Python still feels like the most descriptive title now that the scope of the module has been limited so much, and much of the other introductory content will be going in other places. 

There was definitely a point in my own learning curve where I already had anaconda installed, but still would have been searching for "introduction to python" content. 

As long as the comment + long description are accurate, this shouldn't matter too too much, but I think still at least worth considering!

Author: I think you're right, it could be a bit more descriptive... I think I like "Getting started with Python"... let's take one last look at the content before I make that change though, since changing file names messes with the change log.

> title: Introduction to Python

Reviewer: ```suggestion
title: Demystifying Python
```

comment:  This module introduces the Python programming language, explores why Python is useful in research, and describes how to download Python and Jupyter.

> long_description: Python is a versatile programming language that is frequently used for data analysis, machine learning, web development, and more. If you are interested in learning about any of these topics and are new to programming, are joining a team where Python is used, or have just heard about Python and are looking for a general introduction to the language, this module is a good place to start. This is appropriate for someone at the beginner level.

Reviewer: ```suggestion
long_description: Python is a versatile programming language that is frequently used for data analysis, machine learning, web development, and more. If you are interested in using Python (or even just trying it out), and are looking for how to get set up, this module is a good place to start. This is appropriate for someone at the beginner level, including those with no prior knowledge of or experience with Python.
```

suggestion to increase specificity of what exactly is covered in this module, since this description was developed for the bigger previous module. (e.g., this version of the module doesn't really provide a general intro to the language as much as it just discusses the "why" and then gets learners set up to use it). feel free to totally change or reject changing it at all.

Author: I like this suggestion, it's a bit more targeted to the module as it is now.

### Details

**Estimated time to completion**: @estimated_time

> **Pre-requisites**: Learners should be familiar with tabular data (data stored in a rectangular format, like an Excel spreadsheet or a comma separated file). It is helpful to have some familiarity with navigating to specific directories and running programs in the Command Line/Terminal. Learners do not need to have access to Python or Jupyter notebooks on their own computers.

Author: Yep good call!

</div>
***

## Python Code

Python can be written in several ways:

* You can write Python code interactively using your computer's Terminal program or Command Line (if you're not sure where that is or how to use it, the [Bash/Command Line 101 module](https://liascript.github.io/course/?https://raw.githubusercontent.com/arcus/education_modules/main/bash_command_line_101/bash_command_line_101.md#1) will help you get started).
* You can write scripts that include Python code.
* You can create Python notebooks using [Jupyter](https://jupyter.org/) or online using [Google Colaboratory](https://colab.research.google.com/?utm_source=scs-index) (or "Colab" for short).

In most cases, you'll need to start with downloading Python to your computer (the exception is when using cloud-based solutions like Google Colab); which option you choose will come down to personal preference and your specific goals.

If you have Python installed on your computer, you can use your computer's Command Line or Terminal program to write Python code interactively using a Python **interpreter** by typing `python3`:

<div style="display:none">@gifPreload</div>

<figure>

>   <img src="https://github.com/arcus/education_modules/blob/main/intro_to_python/media/python_interactive.png?raw=true" height="500" width="800" alt="An example of Python code in the Terminal with the following code: 7 + 8, followed by an output of 15;  print('Hello World!'), which gives the output 'Hello World!'; and list(range(0,10)), which prints all integers from 0 to 9" data-alt="https://github.com/arcus/education_modules/blob/main/intro_to_python/media/python_interactive.gif?raw=true">

Reviewer: ```suggestion
  <img src="https://github.com/arcus/education_modules/blob/main/demystifying_python/media/python_interactive.png?raw=true" height="500" width="800" alt="An example of Python code in the Terminal with the following code: 7 + 8, followed by an output of 15;  print('Hello World!'), which gives the output 'Hello World!'; and list(range(0,10)), which prints all integers from 0 to 9" data-alt="https://github.com/arcus/education_modules/blob/main/demystifying_python/media/python_interactive.gif?raw=true">
```

In the script above you can also see lines of text marked with a pound sign/hash mark `#` at the beginning. The pound sign/hash mark indicates the start of a **comment**, which is anything you write in a script that you don't want to run as code. Comments are a good way to briefly describe or explain aspects of your script, and can also be used to "hide" lines of code as you experiment (an alternative to deleting code that you may want to add back later).

The Spyder IDE has a console where you can see the output of your script without having to run it in the Command Prompt/Terminal. As you can see in the screenshot below, all we had to do was create the file one time and then run that file instead of typing in the code interactively. This can save you time and effort if you have a task you know you'll want to perform repeatedly. In the Spyder IDE there is a "run" button (a green arrow in menu at the top) but you could also run the same file in your Terminal/Command Line by typing the command `python` followed by the file name of your Python script.

<div style="display:none">@gifPreload</div>

<figure>

>   <img src="https://github.com/arcus/education_modules/blob/main/intro_to_python/media/python_spyder_console.png?raw=true" height="500" width="800" alt="Running the Python script discussed above in Spyder, showing a user typing 'Pythonista' in response to the prompt and showing the output 'Welcome to Intro to Python, Pythonista!'" data-alt="https://github.com/arcus/education_modules/blob/main/intro_to_python/media/python_spyder_console.gif?raw=true">

Reviewer: ```suggestion
  <img src="https://github.com/arcus/education_modules/blob/main/demystifying_python/media/python_spyder_console.png?raw=true" height="500" width="800" alt="Running the Python script discussed above in Spyder, showing a user typing 'Pythonista' in response to the prompt and showing the output 'Welcome to Intro to Python, Pythonista!'" data-alt="https://github.com/arcus/education_modules/blob/main/demystifying_python/media/python_spyder_console.gif?raw=true">
```

You can even see what this module's notebook looks like in Google Colab! Once you've worked through the introduction above:

1. Click "File" in the upper left-hand corner and then "Open notebook" in the drop-down menu:

   ![To open a notebook in Google Colab, click the File drop down menu in the top left, and select the second option: Open notebook](media/google_colab_open_nb.png)

 2. In the window that appears, one of the options at the top is **GitHub**. Click that, and you'll see where you can type in a GitHub URL.
 3. In that space, copy and paste the repository where these modules live: [https://github.com/arcus/education_modules](https://github.com/arcus/education_modules).
 4. Click the magnifying glass to the far right to search. You should see a list of the Python notebooks in that repository.
 5. When you find the one you want, the button on the far right (an upward-slanting arrow in a box) will open the notebook file as a Colab notebook in another tab.

   <div style="display:none">@gifPreload</div>

   <figure>

>      <img src="https://github.com/arcus/education_modules/blob/update-intro-to-python/intro_to_python/media/github_repo_colab.png?raw=true" height="500" width="800" alt="Finding a GitHub repository in Google Colab following the above instructions." data-alt="https://github.com/arcus/education_modules/blob/update-intro-to-python/intro_to_python/media/github_repo_colab.gif?raw=true">

Reviewer: ```suggestion
     <img src="https://github.com/arcus/education_modules/blob/main/demystifying_python/media/github_repo_colab.png?raw=true" height="500" width="800" alt="Finding a GitHub repository in Google Colab following the above instructions." data-alt="https://github.com/arcus/education_modules/blob/main/demystifying_python/media/github_repo_colab.gif?raw=true">
```

   <figcaption>Click on the image to play the demo.</figcaption>
   </figure>

### Quiz: Accessing Python on Your Computer

How can you access Python to begin practicing on your own?

[( )] By downloading the latest version at https://www.python.org/downloads/
[( )] By downloading the latest Anaconda distribution
[( )] By using cloud-based tools, such as Google Colab
[(X)] All of the above
***

<div class = "answer">

You can download Python directly to your computer, use Anaconda to download Python as well as many other data science tools, or use cloud-based notebooks like Google Colab. The choice is yours!

</div>
***

## Additional Resources

> 

Reviewer: add link to other python module that was split out from intro to python? (I'm not sure what the new name is, otherwise I'd add it myself as a suggestion! :) )

[[X]] Python is human-readable.
[[X]] The Python community has lots of resources to help you get started.
***
<div class = "answer">
While Python is human-readable and free, it is not point-and-click software, and it can take some time and effort to learn how to write Python code. Don't worry, there are lots of resources out there to help you learn!
</div>
***

## Python Code

Python can be written in several ways. You can write Python code interactively using your computer's Terminal program or Command Line, you can write scripts that include Python code, or you can create Python notebooks using [Jupyter](https://jupyter.org/) or online using [Google Colaboratory](https://colab.research.google.com/?utm_source=scs-index) (or "Colab" for short). In most cases, you'll need to start with downloading Python to your computer (the exception will be if you are using cloud-based solutions like Google Colab); which option you choose will come down to personal preference and your specific goals.

If you have Python installed on your computer, you can use your computer's Command Line or Terminal program to write Python code interactively using a Python **interpreter** by typing `python3`:

> 

Author: Done!

## Python Code

Python can be written in several ways:
- You can write Python code interactively using your computer's Terminal program or Command Line
- You can write scripts that include Python code
- You can create Python notebooks using [Jupyter](https://jupyter.org/) or online using [Google Colaboratory](https://colab.research.google.com/?utm_source=scs-index) (or "Colab" for short).

In most cases, you'll need to start with downloading Python to your computer (the exception will be if you are using cloud-based solutions like Google Colab); which option you choose will come down to personal preference and your specific goals.

If you have Python installed on your computer, you can use your computer's Command Line or Terminal program to write Python code interactively using a Python **interpreter** by typing `python3`:

<div style="display:none">@gifPreload</div>

<figure>

>   <img src="https://github.com/arcus/education_modules/blob/main/intro_to_python/media/python_interactive.png?raw=true" height="500" width="800" alt="An example of Python code in the Terminal with the following code: 7 + 8, after which the number 15 is displayed as output;  print('Hello World!'), which gives the output 'Hello World!'; and list(range(0,10)), which gives the output [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]" data-alt="https://github.com/arcus/education_modules/blob/main/intro_to_python/media/python_interactive.gif?raw=true">

Author: Oooh yes much better.

<figcaption>Click on the image to play the demo.</figcaption>
</figure>

To exit the Python interpreter, type the command `exit()` and then press the Enter key. Alternatively, you can also close the interpreter window.

> Using Python interactively in the terminal can be useful if you want to quickly test out short pieces of code. Most of the time, though, you'll either be creating Python scripts or using a notebook instead.

Author: Good idea!

> ## Python Scripts

Author: Just to be clear here, you are suggesting having a line break between the the line that ends we "run in order with one command" and the line that begins "Scripts are often written"?

> ## Python Scripts

Author: Ahh, I see, gotcha.

Here is a short script written in the Spyder IDE:
<div style = "margin: 1rem; width: 750;">

> ![Python code.](media/python_in_spyder.png?raw=true)

Author: I like your suggestion, I think it's more descriptive without being overly long and repeating information.

In the above example, our script asks for input of a user's name using the `input()` function, stores that input in a variable we've called `name`, and prints out a welcome message to the screen that includes the name the user entered using the code `print("Welcome to Intro to Python, "+name+"!")`.

> In the script above you can also see lines of text marked with a pound sign/hash mark `#` at the beginning. The pound sign/hash mark indicates the start of a **comment**, which is anything you write in a script that you don't want to run as code. Comments are a good way to briefly describe or explain aspects of your script, and can also be used to "hide" lines of code as you experiment (an alternative to deleting code that you may want to add back later).

Reviewer: ```suggestion
In the script above you can also see lines of text marked with a pound sign/hash mark `#` at the beginning. This symbol indicates the start of a **comment**, which is anything you write in a script that you don't want to run as code. Comments are a good way to briefly describe or explain aspects of your script, and can also be used to "hide" lines of code as you experiment (an alternative to deleting code that you may want to add back later).
```

> 

Author: Definitely, good call.

### Quiz: Where to write Python code

Which of the following is **not** a recommended way to write Python code?

[( )] in the Python interpreter in the command line.
[(X)] as a script a Google document.
[( )] as a script in an IDE.
[( )] online in a notebook.
***
<div class = "answer">
While it is technically possible to write Python scripts in a Google document, a word processor like Microsoft Word, or a text editor not designed to write code, it is generally not recommended and will make your life harder. There are a variety of text editors that are specifically designed for writing scripts. They generally support multiple languages and have special features that help you write code, such as syntax highlighting, smart code completion, and collaboration tools. Which features you have access to depends on on which editor you choose ([Atom](https://atom.io/) and [Visual Studio Code](https://code.visualstudio.com/) are two free options).
</div>
***

## Jupyter Notebooks

> If you are interested in using Python for data analysis, you might want to write Python in a [Jupyter notebook](https://jupyter.org/) (which was previously called an IPython notebook). There are a couple of reasons why you should consider using Jupyter notebooks:

Author: Haha same!

- You can intersperse sections of text, written in Markdown language, that can provide explanations and context for your code. This is especially valuable for data analysis, similar to a scientific notebook.

### Quiz: Jupyter Notebooks

What are the components of a Jupyter notebook? Select all that apply.

[[X]] Code cells
[[X]] Text cells written in Markdown
[[ ]] A console where you can write lines of code interactively
***
<div class = "answer">
Jupyter notebooks contain code cells and text cells. The code cells are written in Python and can contain as much or as little code as you want; the output of the code can be displayed directly beneath the code cell and the code can be edited and re-run at any time. The text cells are written in Markdown language and can provide more context and reasoning than can easily be done using comments. Additionally, using Markdown allows you to add useful formatting to your text. Jupyter notebooks do not have console, but since you can run each code cell independently, you don't really need one!
</div>
***

> ## Accessing Python on Your Computer

Reviewer: ```suggestion
## Accessing Python on Your Computer or in the Cloud
```

```suggestion
## Accessing Python
```
I think either of these would work -- just don't want someone to glance at the TOC and think "on your computer" means running locally and then decide this module isn't right fit because they're trying to learn how to do in the cloud/google colab stuff.

> ## Accessing Python on Your Computer

Reviewer: ```suggestion
## Accessing Python on Your Computer or in the Cloud
```

```suggestion
## Accessing Python
```
I think either of these would work -- just don't want someone to glance at the TOC and think "on your computer" means running locally and then decide this module isn't right fit because they're trying to learn how to do in the cloud/google colab stuff.

=======

If you don't want to download Python and would instead prefer to continue practicing in the cloud, [Google Colaboratory](https://colab.research.google.com/), or Google Colab, provides an online Jupyter notebook environment that is completely free (though you do need a Google account). The introduction is actually written as a Colab notebook itself, so you can jump right in practicing. While it won't look exactly like the Jupyter notebook we used for this module, it will be a very similar experience, with a mix of code cells and text cells written in Markdown.

You can even see what this module's notebook looks like in Google Colab! Once you've worked through the introduction above:

1. Click "File" in the upper left-hand corner and then "Open notebook" in the drop-down menu:

   ![Opening a notebook in Google Colab](media/google_colab_open_nb.png)

 2. In the window that appears, one of the options at the top is **GitHub**. Click that, and you'll see where you can type in a GitHub URL.
 3. In that space, copy and paste the repository where these modules live: [https://github.com/arcus/education_modules](https://github.com/arcus/education_modules).

>  4. Click the magnifying glass to the far right. You should see a list of the Python notebooks in that repository.

Author: Good to know!

 5. When you find the one you want, the button on the far right (an upward-slanting arrow in a box) will open the notebook file as a Colab notebook in another tab.

   <div style="display:none">@gifPreload</div>

   <figure>

>      <img src="https://github.com/arcus/education_modules/blob/update-intro-to-python/intro_to_python/media/github_repo_colab.png?raw=true" height="500" width="800" alt="Finding a GitHub repository in Google Colab" data-alt="https://github.com/arcus/education_modules/blob/update-intro-to-python/intro_to_python/media/github_repo_colab.gif?raw=true">

Reviewer: I'm conflicted here -- since the info is all contained in the text above, this would typically be considered redundant and the alt would then require a blank tag. On the other hand, the fig caption says "click on the image to play the demo", so users w/o visual access are still going to be aware of the presence of an image, and perhaps then having a blank tag could make them wonder if they're missing out on important stuff....

What are your thoughts?

>      <img src="https://github.com/arcus/education_modules/blob/update-intro-to-python/intro_to_python/media/github_repo_colab.png?raw=true" height="500" width="800" alt="Finding a GitHub repository in Google Colab" data-alt="https://github.com/arcus/education_modules/blob/update-intro-to-python/intro_to_python/media/github_repo_colab.gif?raw=true">

Reviewer: I'm conflicted here -- since the info is all contained in the text above, this would typically be considered redundant and the alt would then require a blank tag. On the other hand, the fig caption says "click on the image to play the demo", so users w/o visual access are still going to be aware of the presence of an image, and perhaps then having a blank tag could make them wonder if they're missing out on important stuff....

What are your thoughts?

   <figcaption>Click on the image to play the demo.</figcaption>
   </figure>

### Quiz: Accessing Python on Your Computer

How can you access Python to begin practicing on your own?

[( )] By downloading the latest version at https://www.python.org/downloads/
[( )] By downloading the latest Anaconda distribution
[( )] By using cloud-based tools, such as Google Colab
[(X)] All of the above
***

<div class = "answer">

You can download Python directly to your computer, use Anaconda to download Python as well as many other data science tools, or use cloud-based notebooks like Google Colab. The choice is yours!

</div>
***

## Additional Resources

> * [python.org](https://www.python.org/) is a great resource for documentation, FAQ's, and tutorials for beginners, as well as information about what is happening in the wider Python community. Check it out and explore!

Reviewer: ```suggestion
* [python.org](https://www.python.org/) is a great resource for documentation, FAQs, and tutorials for beginners, as well as information about what is happening in the wider Python community. Check it out and explore!
```
```suggestion
* [python.org](https://www.python.org/) is a great resource for documentation, FAQ's, and tutorials for beginners, as well as information about what is happening in the wider Python community. Check it out and explore!
```

> * [python.org](https://www.python.org/) is a great resource for documentation, FAQ's, and tutorials for beginners, as well as information about what is happening in the wider Python community. Check it out and explore!

Reviewer: ```suggestion
* [python.org](https://www.python.org/) is a great resource for documentation, FAQs, and tutorials for beginners, as well as information about what is happening in the wider Python community. Check it out and explore!
```
```suggestion
* [python.org](https://www.python.org/) is a great resource for documentation, FAQ's, and tutorials for beginners, as well as information about what is happening in the wider Python community. Check it out and explore!
```

You can even see what this module's notebook looks like in Google Colab! Once you've worked through the introduction above:

1. Click "File" in the upper left-hand corner and then "Open notebook" in the drop-down menu:

   ![To open a notebook in Google Colab, click the File drop down menu in the top left, and select the second option: Open notebook](media/google_colab_open_nb.png)

 2. In the window that appears, one of the options at the top is **GitHub**. Click that, and you'll see where you can type in a GitHub URL.
 3. In that space, copy and paste the repository where these modules live: [https://github.com/arcus/education_modules](https://github.com/arcus/education_modules).
 4. Click the magnifying glass to the far right to search. You should see a list of the Python notebooks in that repository.
 5. When you find the one you want, the button on the far right (an upward-slanting arrow in a box) will open the notebook file as a Colab notebook in another tab.

   <div style="display:none">@gifPreload</div>

   <figure>

>      <img src="https://github.com/arcus/education_modules/blob/update-intro-to-python/intro_to_python/media/github_repo_colab.png?raw=true" height="500" width="800" alt="Finding a GitHub repository in Google Colab following the above instructions." data-alt="https://github.com/arcus/education_modules/blob/update-intro-to-python/intro_to_python/media/github_repo_colab.gif?raw=true">

Reviewer: this seems like a sufficient compromise between the conflict of blank alt text tag vs knowledge of an image and lack of certainty of if there's any missing info. 

And of course, if we ever get feedback from folks without visual access, we can adjust based on that!
